doc/cprover-manual/jbmc-user-manual.md

```diff
@@ -0,0 +1,229 @@
+\ingroup module_hidden
+\page jbmc-user-manual JBMC User Manual
+
+
+\section jbmc-manual-intro Introduction
+
+JBMC is a bounded model checking tool for the Java language. It follows the
+same principles as CBMC, thus the existing
+[documentation for CBMC](http://www.cprover.org/cprover-manual/cbmc.html)
+largely applies to JBMC.
+
+In brief, JBMC creates a model from Java Bytecode (given as input as .class
+files) and performs static analysis and symbolic simulation which allow to
+prove or refute properties. These properties are either automatically
+generated, or provided by the user in the form of assertions.
+
+Using examples, the following section will guide you through the basic
+features of JBMC, and explore the most useful features.
+
+
+\section jbmc-tutorial JBMC tutorial
+
+\subsection jbmc-manual-auto-assert Automatic assertions
+
+Consider the following simplistic `Example.java` file which we have compiled
+into `Example.class` and which contains a utility method to query a given
+array at a given index:
+```java
+1  package core;
+2
+3  public class Example {
+4
+5      public static int query(int[] array, int index) {
+6          return array[index];
+7      }
+8  }
+```
+
+Now let's run the following command to let JBMC tell us about potential errors
+in our `query` method.
+```
+jbmc Example.class --function core.Example.query
+```
+
+The output contains the following:
+```
+core/Example.java function java::core.Example.query:([II)I
+[java::core.Example.query:([II)I.null-pointer-exception.1] line 6 Null pointer check: FAILURE
+[java::core.Example.query:([II)I.array-index-out-of-bounds-low.1] line 6 Array index should be >= 0: FAILURE
+[java::core.Example.query:([II)I.array-index-out-of-bounds-high.1] line 6 Array index should be < length: FAILURE
+[java::core.Example.query:([II)I.1] line 6 no uncaught exception: SUCCESS
+```
+Three reported failures spring up:
+1) there is no null pointer check on the array passed as argument
+2) there is no check that the index at which we access the array is non-negative
+3) there is no check that the index is within the length of the array.
+
+If we make our code more robust by adding guards to make the array access
+safe as follows:
+```java
+5      public static int query(int[] array, int index) {
```

Comment:
I suggest to make the examples complete classes with different names so that the user can easily copy paste the code and run.

Reply:
Done